server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -145,6 +146,20 @@ protected BulkShardResponse newResponseInstance(StreamInput in) throws IOExcepti
         return new BulkShardResponse(in);
     }
 
+    @Override
+    protected void shardOperationOnPrimary(
+        BulkShardRequest request,
+        IndexShard primary,
+        ActionListener<PrimaryResult<BulkShardRequest, BulkShardResponse>> listener
+    ) {
+        final ActionListener<Void> wrappedListener = listener.delegateFailure(
+            (l, ignored) -> super.shardOperationOnPrimary(request, primary, l)
+        );
+        try (var preBulkProceedListeners = new RefCountingListener(wrappedListener)) {
```

Comment:
This is in a really hot code path, do you think that it'll have some performance impact?

Reply:
Not at all, it's just a CAS for regular unhollow shards. And for hollow shards, it'll simply add a listener and trigger unhollowing (on another thread pool).